[WIP][FX] Fix empty tuple type_repr

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58162 [WIP][FX] Fix empty tuple type_repr**

Previously for empty tuple type annotations, `type_repr` was emitting `typing_Tuple[_]`, which is bizarre and not a correct syntax for an empty tuple annotation. This makes it so that it prints `typing.Tuple[()]`, which is the correct syntax.


NOTE: there seems to be an issue with TorchScript parsing an empty tuple with a mangled type name. For example, this works:

```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[int]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
scripted = torch.jit.script(traced)
```

But this fails:
```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[()]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
print(traced.code)
scripted = torch.jit.script(traced)
```

```
RuntimeError: 
Unknown type constructor typing_Tuple:
def forward(self, x : typing_Tuple[()]):
                      ~~~~~~~~~~~~~ <--- HERE
    getitem = x[0];  x = None
    return getitem
```

Differential Revision: [D28386267](https://our.internmc.facebook.com/intern/diff/D28386267)

Commit:
[FX] Fix empty tuple type_repr

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28386267).

Commit:
Update on "[WIP][FX] Fix empty tuple type_repr"


Previously for empty tuple type annotations, `type_repr` was emitting `typing_Tuple[_]`, which is bizarre and not a correct syntax for an empty tuple annotation. This makes it so that it prints `typing.Tuple[()]`, which is the correct syntax.


NOTE: there seems to be an issue with TorchScript parsing an empty tuple with a mangled type name. For example, this works:

```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[int]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
scripted = torch.jit.script(traced)
```

But this fails:
```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[()]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
print(traced.code)
scripted = torch.jit.script(traced)
```

```
RuntimeError: 
Unknown type constructor typing_Tuple:
def forward(self, x : typing_Tuple[()]):
                      ~~~~~~~~~~~~~ <--- HERE
    getitem = x[0];  x = None
    return getitem
```

Differential Revision: [D28386267](https://our.internmc.facebook.com/intern/diff/D28386267)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix empty tuple type_repr"


Previously for empty tuple type annotations, `type_repr` was emitting `typing_Tuple[_]`, which is bizarre and not a correct syntax for an empty tuple annotation. This makes it so that it prints `typing.Tuple[()]`, which is the correct syntax.


NOTE: there seems to be an issue with TorchScript parsing an empty tuple with a mangled type name. For example, this works:

```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[int]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
scripted = torch.jit.script(traced)
```

But this fails:
```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[()]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
print(traced.code)
scripted = torch.jit.script(traced)
```

```
RuntimeError: 
Unknown type constructor typing_Tuple:
def forward(self, x : typing_Tuple[()]):
                      ~~~~~~~~~~~~~ <--- HERE
    getitem = x[0];  x = None
    return getitem
```

Differential Revision: [D28386267](https://our.internmc.facebook.com/intern/diff/D28386267)

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -24,8 +24,12 @@ std::string collectQualname(const Select& select) {
+  if (typeName.rfind("typing.", 0) == 0) {
+    size_t pos = 7; // length of "typing."
+    typeName = typeName.substr(pos);
+  }
```
@ansley I don't think this is the right approach. The first example in the PR description shows that the script type parser (or some component upstream) can already resolve `typing_List` and `typing_Tuple` fine, the only issue arises when we specifically see `typing_Tuple[()]`. I'll dig in to see how those aliases are resolved currently

jamesr66a replied:
@ansley the issue appears to be in the resolution callback stuff here:

https://github.com/pytorch/pytorch/blob/9bfc1c4e0ee68b07557dd1138b3a9a6a587fa8a1/torch/_jit_internal.py#L73


When it encountered the expression `typing_Tuple[()]`, the parsing rules seem to diverge from Python and it discards the `()` tuple literal. Then it tries to lookup an empty string on the module, returning None

jamesr66a replied:
question in my mind: is there a reason we can't use script_type_parser there?


or alternatively, eval the expression in the module environment then throw ann_to_type at it: https://github.com/pytorch/pytorch/blob/9bfc1c4e0ee68b07557dd1138b3a9a6a587fa8a1/torch/jit/annotations.py#L295

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28386267).

Commit:
Update on "[WIP][FX] Fix empty tuple type_repr"


Previously for empty tuple type annotations, `type_repr` was emitting `typing_Tuple[_]`, which is bizarre and not a correct syntax for an empty tuple annotation. This makes it so that it prints `typing.Tuple[()]`, which is the correct syntax.


NOTE: there seems to be an issue with TorchScript parsing an empty tuple with a mangled type name. For example, this works:

```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[int]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
scripted = torch.jit.script(traced)
```

But this fails:
```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[()]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
print(traced.code)
scripted = torch.jit.script(traced)
```

```
RuntimeError: 
Unknown type constructor typing_Tuple:
def forward(self, x : typing_Tuple[()]):
                      ~~~~~~~~~~~~~ <--- HERE
    getitem = x[0];  x = None
    return getitem
```

Differential Revision: [D28386267](https://our.internmc.facebook.com/intern/diff/D28386267)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28386267).

Commit:
Update on "[WIP][FX] Fix empty tuple type_repr"


Previously for empty tuple type annotations, `type_repr` was emitting `typing_Tuple[_]`, which is bizarre and not a correct syntax for an empty tuple annotation. This makes it so that it prints `typing.Tuple[()]`, which is the correct syntax.


NOTE: there seems to be an issue with TorchScript parsing an empty tuple with a mangled type name. For example, this works:

```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[int]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
scripted = torch.jit.script(traced)
```

But this fails:
```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[()]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
print(traced.code)
scripted = torch.jit.script(traced)
```

```
RuntimeError: 
Unknown type constructor typing_Tuple:
def forward(self, x : typing_Tuple[()]):
                      ~~~~~~~~~~~~~ <--- HERE
    getitem = x[0];  x = None
    return getitem
```

Differential Revision: [D28386267](https://our.internmc.facebook.com/intern/diff/D28386267)

[ghstack-poisoned]

Commit:
Update on "[WIP][FX] Fix empty tuple type_repr"


Previously for empty tuple type annotations, `type_repr` was emitting `typing_Tuple[_]`, which is bizarre and not a correct syntax for an empty tuple annotation. This makes it so that it prints `typing.Tuple[()]`, which is the correct syntax.


NOTE: there seems to be an issue with TorchScript parsing an empty tuple with a mangled type name. For example, this works:

```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[int]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
scripted = torch.jit.script(traced)
```

But this fails:
```
import torch
import torch.fx
from typing import Tuple, List

class Foo(torch.nn.Module):
    def forward(self, x : Tuple[()]):
        return x[0]

traced = torch.fx.symbolic_trace(Foo())
print(traced.code)
scripted = torch.jit.script(traced)
```

```
RuntimeError: 
Unknown type constructor typing_Tuple:
def forward(self, x : typing_Tuple[()]):
                      ~~~~~~~~~~~~~ <--- HERE
    getitem = x[0];  x = None
    return getitem
```

Differential Revision: [D28386267](https://our.internmc.facebook.com/intern/diff/D28386267)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28386267).

jamesr66a commented on torch/jit/annotations.py:
```diff
@@ -300,6 +300,8 @@ def try_ann_to_type(ann, loc):
     if inspect.isclass(ann) and is_tensor(ann):
         return TensorType.get()
     if is_tuple(ann):
+        if len(ann.__args__) == 1 and ann.__args__[0] == ():
```
ubernit: does `if ann.__args__ == [()]:` work?

ansley replied:
surprisingly not!! Future PEP idea? :(

jamesr66a commented on torch/_jit_internal.py:
```diff
@@ -60,6 +60,9 @@ def parseNestedExpr(expr, module) -> Tuple[Any, int]:
             i += 1
 
+        if expr[:i] == '()':
+            return (), i
```
Add a comment to explain what's going on :)

jamesr66a commented on torch/jit/annotations.py:
```diff
@@ -300,6 +300,8 @@ def try_ann_to_type(ann, loc):
         return TensorType.get()
     if is_tuple(ann):
+        if len(ann.__args__) == 1 and ann.__args__[0] == ():
+            return TupleType([])
```
Also add a comment here. It's probably a good idea to comment wherever this empty tuple special casing happens, since it's unintuitive behavior of python. Linking to the definition of the type annotations for tuple, which discusses this could help https://docs.python.org/3/library/typing.html#typing.Tuple

jamesr66a reviewed (commented):
@ansley  looks good to me, maybe you should copy out the changes into your own PR and I can approve it

ansley commented:
We're moving this to [a different branch](https://github.com/pytorch/pytorch/pull/58340) since it's confusing to have James reviewing "his" PR
